### ISSUE TITLE
Allow setting existing parameters from the command line

### ISSUE THREAD

**Issue by vkarak** [reframe-hpc/reframe#2674]
That's a current limitation. You cannot reset a parameter from the command line. It would be nice to be able to do:

```python
class my_test(...):
    x = parameter(range(10))
```

```
reframe --parametrise my_test.x=1,4,5
```

and change the parameterisation of `my_test`.

**Comment by casparvl** [comment]
+1, we'd be interested in that. I'd like to have e.g. a test that by default parametrizes over all `GROMACS` modules find by `find_modules`. But sometimes, I may want to run the test just with _one particular_ `GROMACS`. It would be great if we can then overwrite on command line like this.

**Comment by ekouts** [comment]
> But sometimes, I may want to run the test just with one particular GROMACS

The parameterize option that @vkarak suggests would still be useful if you the new values are not included in the test's parameters, but for many cases selecting the test you like with the `-n` option may be enough, right?

**Comment by vkarak** [comment]
Yes, the idea is to be able to override the parameter values completely. That's a superset of what @casparvl requests, but it would still make his scenario much faster if we supported this option. :-) Currently, also what you suggest @ekouts would work fine, and it's much more convenient now with the test hashes as that you can quickly select a test with `-n /<hash>`. But you still have to pay the cost of creating all the test variants, which can be significant using the `find_modules()` that interacts with the modules system.

**Comment by mredenti** [comment]
I was looking into a couple of options for this. 

**Option 1: Extend the existing `-S`/`--setvar` option**

Currently this option allows to ovverride test variables and this behaviour could me modified to also support setting parameter values. 

Discussion Points
User Experience: How intuitive and easy to use is the option for end-users?
Implementation Complexity: How complex is the implementation, and how much does it affect the existing codebase?
Flexibility: Does the option provide the necessary flexibility to cover all use cases?
Recommendation
Based on the considerations, Option 2: Extend the Existing --setvar Option is recommended. This approach provides a clear and intuitive way to override parameter values without affecting the existing --parameterize logic. It also keeps the implementation relatively straightforward.

**Option 1: Extend the existing `--parameterize` option**

The current behaviour with the `-P`/`--parameterize` cl argument is to allow the parameterization of a test on an existing `variable()`. So, one approach could be to extend this option to support overriding also `parameter()` values. 

```python
testgen_options.add_argument(
        '-P', '--parameterize', action='append', metavar='VAR:VAL0,VAL1,... | PAR:VAL0,VAL1,...',
        default=[], help='Parameterize a test on a set of variables or parameters'
    )
```

The advantage would be the reuse of the existing `--parameterize` option, keeping the CLI interface simpler and probably the parsing login would not be to challenging to modify. But variables and parameters have a different role and perhaps this cross behaviour is a bit confusing. 

**Option 2: Change the behaviour of the existing `--parameterize` option**

Variables are variables and parameters are parameters. So the `-P`/`--parameterize` option should only deal with parameters. 

Option 2: Introduce a New --override-param Option
Description: Introduce a new option, --override-param, specifically for overriding existing parameter values. This option would work similarly to --parameterize but would replace the existing values instead of adding new ones.

Pros:

Clear separation of functionality between parameterizing and overriding.
Easier to implement without affecting the existing --parameterize logic.
Cons:

Adds another option to the CLI, which may increase complexity for users.
Users need to learn and remember the new option.
Example Usage:

Discussion Points
User Experience: How intuitive and easy to use is the option for end-users?
Implementation Complexity: How complex is the implementation, and how much does it affect the existing codebase?
Flexibility: Does the option provide the necessary flexibility to cover all use cases?

**Comment by mredenti** [comment]
I was looking into a couple of options for this.  I think the most suitable option is probably the following.

**Extend the existing `-S`/`--setvar` option**

Currently this option allows to ovverride test `variable()`s and this behaviour could me modified to also support overriding existing `parameter()` values. 

If we extend the existing `-S`/`--setvar`, it would make sense to rename the option to something more generic that reflects its broader functionality.  Probably `--set` instead of `--setvar` 

```python
run_options.add_argument(
    '-S', '--set', action='append', metavar='[TEST.]VAR=VAL | [TEST.]PARAM=VAL',
    dest='vars', default=[],
    help=('Set test variable VAR or parameter PARAM to VAL in all tests '
          'or optionally in TEST only')
)
```

The only disadvantage I see with this approach is that `variable()`s and `parameter()`s are truly separate entities in Reframe, so perhaps that should be reflected in the CLI. On the other hand we already have `--parametrize` which turns a variable into a parameter so maybe it is consistent with the current login

--- 


I have considered also other options like

- **Extend the existing `--parameterize` option** 
  The current behaviour with the `-P`/`--parameterize` cl argument is to allow the parameterization of a test on an existing 
  `variable()`. So, one approach could be to extend this option to support overriding also `parameter()` values. But I think it 
  would be a bit confusing cause `--parameterize` suggests that it is turning something that is not a parameter, in this case a 
  variable, into a parameter. I would discard this approach as it would affect the existing `--parameterize` logic.

- **Add an `--override-param` argument**
  This is something that could be achieving with `-S` without introducing a new cl argument

**Comment by mredenti** [comment]
> That's a current limitation. You cannot reset a parameter from the command line. It would be nice to be able to do:
> 
> class my_test(...):
>     x = parameter(range(10))
> ```
> reframe --parametrise my_test.x=1,4,5
> ```
> 
> and change the parameterisation of `my_test`.

The current behaviour with `--parameterize` is that only once test variants have been generated (https://github.com/reframe-hpc/reframe/blob/85edad2554201158afe15fa108d89d23d51a96e9/reframe/frontend/cli.py#L1246-L1251) the `--parameterize` option is considered (https://github.com/reframe-hpc/reframe/blob/85edad2554201158afe15fa108d89d23d51a96e9/reframe/frontend/cli.py#L1339-L1353). The current behaviour enables to parameterize an existing variable but if we wanted to extend its behaviour to allow overriding exising `parameter()` values - without paying the cost of first creating all the test variants - we would need to move the parameterise logic during the instantation of the tests (https://github.com/reframe-hpc/reframe/blob/85edad2554201158afe15fa108d89d23d51a96e9/reframe/frontend/loader.py#L143-L2480. 

An alternative would be to instead ovverride the behaviour of the `-S`/`--setvar` CL argument to also allow setting the values of `parameter()`s. If we extend the existing `-S`/`--setvar` option to also support overriding `parameter()` values, it would make sense to rename the option to something more generic that reflects its broader functionality, perhaps 

```python
run_options.add_argument(
    '-S', '--set', action='append', metavar='[TEST.]VAR=VAL | [TEST.]PARAM=VAL',
    dest='vars', default=[],
    help=('Set test variable VAR or parameter PARAM to VAL in all tests '
          'or optionally in TEST only')
)
```

**Comment by vkarak** [comment]
> The current behaviour enables to parameterize an existing variable but if we wanted to extend its behaviour to allow overriding exising parameter() values - without paying the cost of first creating all the test variants - we would need to move the parameterise logic during the instantation of the tests

This is not accurate. The variables are set *before* the test is instantiated and precisely here: https://github.com/reframe-hpc/reframe/blob/16f17462ea7d6b53ac26f2dff7f4f863d130920c/reframe/core/meta.py#L544

It's during the "instantation" process in `instantiate_all()` but happens just before the actual instantiation of the test class. In fact, setting a variable the exact equivalent of adding a line `var = val` at the end of the class body.

**Comment by mredenti** [comment]
Ok I see. So now the desired behaviour of `--parameterize` would be to parameterise an existing variable **and** override the values of an existing `parameter()`? Perhaps through a `setparam` method equivalent to `setvar`?
